ta/bti_test/ta_entry.c

```diff
@@ -0,0 +1,99 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Linaro Limited
+ * All rights reserved.
+ */
+
+#include <inttypes.h>
+#include <string.h>
+#include <ta_arm_bti.h>
+#include <ta_arm_bti_priv.h>
+#include <tee_internal_api.h>
+
+/*
+ * Trusted Application Entry Points
+ */
+
+/* Called each time a new instance is created */
+TEE_Result TA_CreateEntryPoint(void)
+{
+	return TEE_SUCCESS;
+}
+
+/* Called each time an instance is destroyed */
+void TA_DestroyEntryPoint(void)
+{
+}
+
+/* Called each time a session is opened */
+TEE_Result TA_OpenSessionEntryPoint(uint32_t nParamTypes __unused,
+				    TEE_Param pParams[4] __unused,
+				    void **ppSessionContext __unused)
+{
+	return TEE_SUCCESS;
+}
+
+/* Called each time a session is closed */
+void TA_CloseSessionEntryPoint(void *pSessionContext __unused)
+{
+}
+
+static TEE_Result check_bti_implemented(uint32_t param_types,
+					TEE_Param params[4])
+{
+	bool implemented = false;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (param_types !=
+	    TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+			    TEE_PARAM_TYPE_NONE,
+			    TEE_PARAM_TYPE_NONE,
+			    TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	res = TEE_GetPropertyAsBool(
+			TEE_PROPSET_TEE_IMPLEMENTATION,
+			"org.trustedfirmware.optee.cpu.feat_bti_implemented",
+			&implemented);
+	if (res == TEE_SUCCESS && implemented)
+		params[0].value.a = 1;
+
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+		params[0].value.a = 0;
+		res = TEE_SUCCESS;
+	}
+
+	return res;
+}
+
+__weak TEE_Result test_bti(uint32_t nCommandID __unused,
+			   uint32_t nParamTypes __unused,
+			   TEE_Param pParams[4] __unused)
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
```

Comment:
This is what we get when the TA is 32-bits, and `xtest_tee_test_1035()` doesn't expect that. Should it accept both `TEEC_SUCESS` and `TEEC_ERROR_NOT_IMPLEMENTED`?

Reply:
thanks..Good point. I have modified the xtest_tee_test_1035 to skip in case the first call returns TEEC_ERROR_NOT_IMPLEMENTED